### PR TITLE
Increase performance & make it compatible with more browsers

### PR DESCRIPTION
**Pull request by lovasoa**
Switch from using 1 pixel `fillRect` on an offscreen canvas to using a typed array on the workers. This improves rendering performance, and makes this work on non-chrome browsers.
Tested on chrome, firefox and safari.

### CODE REVIEW

File: js/compy-stuff.js [hunk 1]
```diff
@@ -35,7 +35,7 @@ function testMandelbrot(c, maxIter = 200) {
 
         z = complexAdd(complexMult(z, z), c)
 
-        if (complexAbs(z) > 2) {
+        if (squareComplexAbs(z) > 4) {
```
Comment (BadgerBadgerBadgerBadger) [inline js/compy-stuff.js]: @lovasoa Neat. Didn't know you could do this.

File: js/painter.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 let ctx
 let drawing = false
 
-const numOfAcolytes = 4
+const numOfAcolytes =  navigator.hardwareConcurrency
```
Comment (BadgerBadgerBadgerBadger) [inline js/painter.js]: @lovasoa TIL, again!

File: js/sketch.js [hunk 3]
```diff
@@ -1,11 +1,10 @@
-let canvas
-const width = 600
-const height = 450
+let canvas = document.getElementById('fractal')
+const { width, height } = canvas.getBoundingClientRect();
+const xBounds = { min: -3, max: 1 }
+const yBounds = { min: -1.5, max: 1.5 }
+yBounds.max = yBounds.min + (xBounds.max - xBounds.min) * height / width;
```
Comment (BadgerBadgerBadgerBadger) [inline js/sketch.js]: @lovasoa Looks like this is doing making sure the aspect ratio works out, yeah?

File: js/painter.js [hunk 4]
```diff
@@ -60,43 +59,57 @@ function onAcolyteMessage(event) {
     }
 }
 
-function painter_draw(data) {
-    const mousePos = {
-        x: _.get(data, 'mousePos.x'),
-        y: _.get(data, 'mousePos.y')
+
+const update_bounds = _.throttle(function update_bounds(bounds) {
```
Comment (BadgerBadgerBadgerBadger) [inline js/painter.js]: @lovasoa Might a request camel-case naming? I'm following the same in the rest of the project.

File: js/compy-stuff.js [hunk 5]
```diff
@@ -11,15 +11,36 @@
...
+        }
     }
     return maxIter;
 }
 
+/**
+ * @param {Complex} c
+ * @returns {boolean}
+ */
+function testCardioid(c) {
```
Comment (BadgerBadgerBadgerBadger) [inline js/compy-stuff.js:44]: @lovasoa Could you explain this bit to me? It goes right over my head.

**Review by BadgerBadgerBadgerBadger — COMMENTED** [review]
@lovasoa Thank you so much for the contribution! I'm gonna merge these and then make a few stylistic changes, hope you don't mind. Could you explain some of the improvements you made? They go right over my head. I wanna add an updated section to the blog post explaining your changes and how they help performance.

File: js/compy-stuff.js (hunk 5 above)
Comment (lovasoa) [reply js/compy-stuff.js:44]: `testCardioid` tests whether a point c is in the central cardioid, and `testBulb` tests whether a point is in the circle of center (-1, 0) and radius 1/4. In both cases, we know that the point will be part of the mandelbrot set, and we can save some computation.

image

[More about this on wikipedia](https://en.wikipedia.org/wiki/Mandelbrot_set#Cardioid_/_bulb_checking)

Comment (BadgerBadgerBadgerBadger) [reply js/compy-stuff.js:44]: @lovasoa Thank you!

File: js/sketch.js (hunk 3 above)
Comment (lovasoa) [reply js/sketch.js]: Yes, exactly. This is to make sure the aspect ratio is always right

File: js/compy-stuff.js (hunk 1 above)
Comment (lovasoa) [reply js/compy-stuff.js]: `|z| < 2` <=> `|z|^2 < 4`. It does not change much, but it's always nice when you can avoid a square root.

File: js/acolyte.js [hunk 6]
```diff
@@ -17,57 +14,80 @@ self.onmessage = function onmessage(event) {
...
-    }
-
-    drawing = true
+    const c = new Complex(0, 0);
 
-    for (let x = 0; x < canvas.width; x++) {
-        for (let y = 0; y < canvas.height; y++) {
+    for (let i = 0; i < data.length; i += 4) {
+        let x = offset.x + (i / 4) % dimensions.width;
+        let y = offset.y + (i / 4) / dimensions.width | 0;
```
Comment (BadgerBadgerBadgerBadger) [inline js/acolyte.js:36]: @lovasoa Forgot to ask earlier. What does that bitwise OR do?

Comment (lovasoa) [reply js/acolyte.js:36]: It does the same as `Math.floor`, but it's shorter :) It was the standard way of using integer variables in [asm.js](https://en.wikipedia.org/wiki/Asm.js)

Comment (BadgerBadgerBadgerBadger) [reply js/acolyte.js:36]: @lovasoa Gotcha.

**Pull request merged**